File: sdk/go/vectordb/multimodal.go
package vectordb

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"strings"

	"net/url"
)

// IngestTextParams configures text ingestion.
type IngestTextParams struct {
	Text     string
	Metadata map[string]interface{}
	VectorID string
}

// IngestFileParams configures file-based ingestion (image/audio).
type IngestFileParams struct {
	Contents io.Reader
	Filename string
	Metadata map[string]interface{}
	VectorID string
}

// SearchTextParams configures a text-based search.
type SearchTextParams struct {
	Query   string
	K       int
	Method  string
	Filters map[string]interface{}
}

// SearchFileParams configures a file-based search (image/audio).
type SearchFileParams struct {
	Contents io.Reader
	Filename string
	K        int
	Method   string
	Filters  map[string]interface{}
}

// MultimodalAPI provides access to multimodal (text, image, audio) operations.
type MultimodalAPI struct {
	client *VectorDBClient
}

// IngestText ingests text into a collection.
func (a *MultimodalAPI) IngestText(ctx context.Context, collectionID string, params IngestTextParams, opts ...RequestOption) (map[string]interface{}, error) {
	body := map[string]interface{}{
		"text": params.Text,
	}
	if params.Metadata != nil {
		body["metadata"] = params.Metadata
	}
	if params.VectorID != "" {
		body["vector_id"] = params.VectorID
	}

	o := &requestOptions{}
	for _, fn := range opts {
		fn(o)
	}
	merged := a.client.mergeOpts(o)
	return jsonPost(ctx, a.client.httpClient, a.client.BaseURL, "/collections/"+url.PathEscape(collectionID)+"/ingest/text", body, merged)
}

// SearchText performs a text-based search on a collection and returns typed results.
func (a *MultimodalAPI) SearchText(ctx context.Context, collectionID string, params SearchTextParams, opts ...RequestOption) (SearchResult, error) {
	body := map[string]interface{}{
		"query":  params.Query,
		"k":      ifPositive(params.K, 5),
		"method": orDefault(params.Method, "brute"),
	}
	if params.Filters != nil {
		body["filters"] = params.Filters
	}

	o := &requestOptions{}
	for _, fn := range opts {
		fn(o)
	}
	merged := a.client.mergeOpts(o)

	rr, err := jsonPostRaw(ctx, a.client.httpClient, a.client.BaseURL, "/collections/"+url.PathEscape(collectionID)+"/search/text", body, merged)
	if err != nil {
		return SearchResult{}, err
	}
	return searchResultFromRaw(rr)
}

// IngestImage ingests an image file into a collection.
func (a *MultimodalAPI) IngestImage(ctx context.Context, collectionID string, params IngestFileParams, opts ...RequestOption) (map[string]interface{}, error) {
	if params.Contents == nil {
		return nil, &VectorDBError{msg: "provide contents for image ingest"}
	}
	return a.multipartUpload(ctx, collectionID, "ingest", "image", params, opts...)
}

// SearchImage performs an image-based search and returns typed results.
func (a *MultimodalAPI) SearchImage(ctx context.Context, collectionID string, params SearchFileParams, opts ...RequestOption) (SearchResult, error) {
	if params.Contents == nil {
		return SearchResult{}, &VectorDBError{msg: "provide contents for image search"}
	}
	return a.multipartSearch(ctx, collectionID, "search", "image", params, opts...)
}

// IngestAudio ingests an audio file into a collection.
func (a *MultimodalAPI) IngestAudio(ctx context.Context, collectionID string, params IngestFileParams, opts ...RequestOption) (map[string]interface{}, error) {
	if params.Contents == nil {
		return nil, &VectorDBError{msg: "provide contents for audio ingest"}
	}
	return a.multipartUpload(ctx, collectionID, "ingest", "audio", params, opts...)
}

// SearchAudio performs an audio-based search and returns typed results.
func (a *MultimodalAPI) SearchAudio(ctx context.Context, collectionID string, params SearchFileParams, opts ...RequestOption) (SearchResult, error) {
	if params.Contents == nil {
		return SearchResult{}, &VectorDBError{msg: "provide contents for audio search"}
	}
	return a.multipartSearch(ctx, collectionID, "search", "audio", params, opts...)
}

// Query performs a RAG query on a collection.
func (a *MultimodalAPI) Query(ctx context.Context, collectionID, query string, k int, opts ...RequestOption) (map[string]interface{}, error) {
	body := map[string]interface{}{
		"query": query,
	}
	if k > 0 {
		body["k"] = k
	}

	o := &requestOptions{}
	for _, fn := range opts {
		fn(o)
	}
	merged := a.client.mergeOpts(o)
	return jsonPost(ctx, a.client.httpClient, a.client.BaseURL, "/collections/"+url.PathEscape(collectionID)+"/query", body, merged)
}

// --- Internal helpers ---

func (a *MultimodalAPI) multipartUpload(ctx context.Context, collectionID, action, mediaType string, params IngestFileParams, opts ...RequestOption) (map[string]interface{}, error) {
	path := fmt.Sprintf("/collections/%s/%s/%s", url.PathEscape(collectionID), action, mediaType)
	filename := params.Filename
	if filename == "" {
		filename = "file"
	}

	fields := make(map[string]io.Reader)
	if params.Metadata != nil {
		metaBytes, _ := json.Marshal(params.Metadata)
		fields["metadata"] = strings.NewReader(string(metaBytes))
	}
	if params.VectorID != "" {
		fields["vector_id"] = strings.NewReader(params.VectorID)
	}

	o := &requestOptions{}
	for _, fn := range opts {
		fn(o)
	}
	merged := a.client.mergeOpts(o)

	return multipartPost(ctx, a.client.httpClient, a.client.BaseURL, path, fields, "file", params.Contents, filename, merged)
}

func (a *MultimodalAPI) multipartSearch(ctx context.Context, collectionID, action, mediaType string, params SearchFileParams, opts ...RequestOption) (SearchResult, error) {
	path := fmt.Sprintf("/collections/%s/%s/%s", url.PathEscape(collectionID), action, mediaType)
	filename := params.Filename
	if filename == "" {
		filename = "file"
	}

	fields := make(map[string]io.Reader)
	k := params.K
	if k <= 0 {
		k = 5
	}
	fields["k"] = strings.NewReader(fmt.Sprintf("%d", k))
	if params.Method != "" {
		fields["method"] = strings.NewReader(params.Method)
	}
	if params.Filters != nil {
		filterBytes, _ := json.Marshal(params.Filters)
		fields["filters"] = bytes.NewReader(filterBytes)
	}

	o := &requestOptions{}
	for _, fn := range opts {
		fn(o)
	}
	merged := a.client.mergeOpts(o)

	rr, err := multipartPostRaw(ctx, a.client.httpClient, a.client.BaseURL, path, fields, "file", params.Contents, filename, merged)
	if err != nil {
		return SearchResult{}, err
	}
	return searchResultFromRaw(rr)
}

// searchResultFromRaw unmarshals a search result directly from raw JSON bytes.
func searchResultFromRaw(rr rawResponse) (SearchResult, error) {
	var data SearchResultData
	if err := json.Unmarshal(rr.raw, &data); err != nil {
		return SearchResult{}, err
	}

	hits := make([]SearchHit, len(data.Results))
	for i, r := range data.Results {
		meta := r.Metadata
		if meta == nil {
			meta = r.MetaData
		}
		hits[i] = SearchHit{
			VectorID: r.VectorID,
			Distance: r.Distance,
			Metadata: meta,
		}
	}

	total := len(hits)
	if data.TotalResults != nil {
		total = *data.TotalResults
	}
	var st float64
	if data.SearchTime != nil {
		st = *data.SearchTime
	}

	return SearchResult{
		Success:      data.Success,
		Results:      hits,
		TotalResults: total,
		SearchTime:   st,
		Method:       data.Method,
		Raw:          rr.parsed,
	}, nil
}

File: sdk/typescript/src/multimodal.ts
import type { VectorDBClient } from "./client";
import type { RequestOptions } from "./http";
import { jsonPost, multipartPost } from "./http";
import { SearchResult, type SearchResultData } from "./models";

export interface IngestTextParams {
  text: string;
  metadata?: Record<string, unknown>;
  vectorId?: string;
}

export interface IngestFileParams {
  /** Path (Node.js only — Blob/File in browser) */
  blob?: Blob | File;
  /** Optional metadata encoded as JSON string in the form */
  metadata?: Record<string, unknown>;
  vectorId?: string;
  filename?: string;
}

export interface SearchTextParams {
  query: string;
  k?: number;
  method?: string;
  filters?: Record<string, unknown>;
}

export interface SearchFileParams {
  blob?: Blob | File;
  k?: number;
  method?: string;
  filters?: Record<string, unknown>;
}

export class MultimodalAPI {
  constructor(private readonly client: VectorDBClient) {}

  // ── Text ────────────────────────────────────────────────────────────

  async ingestText(
    collectionId: string,
    params: IngestTextParams,
    options?: RequestOptions,
  ): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { text: params.text };
    if (params.metadata) body.metadata = params.metadata;
    if (params.vectorId) body.vector_id = params.vectorId;
    return jsonPost(
      this.client.baseUrl,
      `/collections/${encodeURIComponent(collectionId)}/ingest/text`,
      body,
      this.client.mergeOptions(options),
    );
  }

  async searchText(
    collectionId: string,
    params: SearchTextParams,
    options?: RequestOptions,
  ): Promise<SearchResult> {
    const body: Record<string, unknown> = {
      query: params.query,
      k: params.k ?? 5,
      method: params.method ?? "brute",
    };
    if (params.filters) body.filters = params.filters;
    const data = await jsonPost(
      this.client.baseUrl,
      `/collections/${encodeURIComponent(collectionId)}/search/text`,
      body,
      this.client.mergeOptions(options),
    );
    return SearchResult.fromApi(data as unknown as SearchResultData);
  }

  // ── Image ───────────────────────────────────────────────────────────

  private makeFileForm(
    blob: Blob | File,
    extra?: { metadata?: Record<string, unknown>; vectorId?: string; k?: number; method?: string; filters?: Record<string, unknown> },
  ): FormData {
    const fd = new FormData();
    fd.append("file", blob);
    if (extra?.metadata) fd.append("metadata", JSON.stringify(extra.metadata));
    if (extra?.vectorId) fd.append("vector_id", extra.vectorId);
    if (extra?.k !== undefined) fd.append("k", String(extra.k));
    if (extra?.method) fd.append("method", extra.method);
    if (extra?.filters) fd.append("filters", JSON.stringify(extra.filters));
    return fd;
  }

  async ingestImage(
    collectionId: string,
    params: IngestFileParams,
    options?: RequestOptions,
  ): Promise<Record<string, unknown>> {
    if (!params.blob) throw new Error("Provide blob for image ingest");
    const fd = this.makeFileForm(params.blob, {
      metadata: params.metadata,
      vectorId: params.vectorId,
    });
    return multipartPost(
      this.client.baseUrl,
      `/collections/${encodeURIComponent(collectionId)}/ingest/image`,
      fd,
      this.client.mergeOptions(options),
    );
  }

  async searchImage(
    collectionId: string,
    params: SearchFileParams,
    options?: RequestOptions,
  ): Promise<SearchResult> {
    if (!params.blob) throw new Error("Provide blob for image search");
    const fd = this.makeFileForm(params.blob, {
      k: params.k ?? 5,
      method: params.method ?? "brute",
      filters: params.filters,
    });
    const data = await multipartPost(
      this.client.baseUrl,
      `/collections/${encodeURIComponent(collectionId)}/search/image`,
      fd,
      this.client.mergeOptions(options),
    );
    return SearchResult.fromApi(data as unknown as SearchResultData);
  }

  // ── Audio ───────────────────────────────────────────────────────────

  async ingestAudio(
    collectionId: string,
    params: IngestFileParams,
    options?: RequestOptions,
  ): Promise<Record<string, unknown>> {
    if (!params.blob) throw new Error("Provide blob for audio ingest");
    const fd = this.makeFileForm(params.blob, {
      metadata: params.metadata,
      vectorId: params.vectorId,
    });
    return multipartPost(
      this.client.baseUrl,
      `/collections/${encodeURIComponent(collectionId)}/ingest/audio`,
      fd,
      this.client.mergeOptions(options),
    );
  }

  async searchAudio(
    collectionId: string,
    params: SearchFileParams,
    options?: RequestOptions,
  ): Promise<SearchResult> {
    if (!params.blob) throw new Error("Provide blob for audio search");
    const fd = this.makeFileForm(params.blob, {
      k: params.k ?? 5,
      method: params.method ?? "brute",
      filters: params.filters,
    });
    const data = await multipartPost(
      this.client.baseUrl,
      `/collections/${encodeURIComponent(collectionId)}/search/audio`,
      fd,
      this.client.mergeOptions(options),
    );
    return SearchResult.fromApi(data as unknown as SearchResultData);
  }

  // ── RAG ─────────────────────────────────────────────────────────────

  async query(
    collectionId: string,
    query: string,
    params?: { k?: number },
    options?: RequestOptions,
  ): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { query };
    if (params?.k) body.k = params.k;
    return jsonPost(
      this.client.baseUrl,
      `/collections/${encodeURIComponent(collectionId)}/query`,
      body,
      this.client.mergeOptions(options),
    );
  }
}
